docs/current/api/975146-concepts.md

```diff
@@ -0,0 +1,126 @@
+
+```graphql
+{
+ "data": {
+  "container": {
+   "from": {
+    "exec": {
+     "stdout": {
+      "contents": "alpine-baselayout-data\nmusl\nbusybox\nalpine-baselayout\nalpine-keys\nca-certificates-bundle\nlibcrypto1.1\nlibssl1.1\nssl_client\nzlib\napk-tools\nscanelf\nmusl-utils\nlibc-utils\n"
+     }
+    }
+   }
+  }
+ }
+}
+```
+
+A GraphQL schema works by defining one or more object types, and then fields (and field arguments) on those object types. The fields of an object type can themselves be objects, allowing for different entities to be logically connected with each other. API users can then perform GraphQL queries to return all or some fields of an object.
+
+With Dagger's GraphQL API, while all of the above is true, each field in a query also resolves to a build operation. To understand this, let's dissect the query above:
+
+1. Consider the first level of the previous query:
+
+  ```graphql
+  query {
+    container {
+      from(address: "alpine:latest") {
+        ...
+      }
+    }
+  }
+  ```
+
+  This query requests the `from` field of Dagger's `Container` object type, passing it the address of a container image as argument. To resolve this, Dagger will initialize a container from the image published at the given address and return a `Container` object representing the new container image.
+
+1. Now, consider the next level of the query:
+
+  ```graphql
+  query {
+    container {
+      from(address: "alpine:latest") {
+        exec(args: ["apk", "info"]) {
+          ...
+        }
+      }
+    }
+  }
+  ```
+
+  Here, the query requests the `exec` field of the `Container` object returned in the previous step, passing it the command to be executed as an array of arguments. To resolve this, Dagger will  define the command for execution in the container image and return a `Container` object containing the execution results.
+
+1. Finally, consider the innermost level of the query:
+
+  ```graphql
+  query {
+    container {
+      from(address: "alpine:latest") {
+        exec(args: ["apk", "info"]) {
+          stdout {
+            contents
+          }
+        }
+      }
+    }
+  }
+  ```
+
+  Here, the query requests the `stdout` field of the `Container` object returned in the previous step. To resolve this, Dagger will return a `File` object type containing the result of the last executed command. The output of the command is retrieved by addressing the `contents` field of the `File` type, which returns a `String`.
+
+## Lazy-Loading
+
+Dagger executes queries "lazily". This means that the individual build operations are performed only when a leaf value (scalar or enum) is requested. In the previous example, Dagger executes the query only when the `stdout` field is requested. This is a key advantage of Dagger: it allows users to write queries (or code, if using an SDK) in a procedural manner, then executes those queries in parallel for maximum speed.
```

Comment:
I read through https://github.com/dagger/dagger/issues/3555, and I understand this might change. I'm assuming it's correct for now, but let me know if not. Also, is there a way for a user to identify in the API reference which fields trigger execution? If not, can you please provide a short list of common ones here?

Reply:
There isn't a list of fields that trigger execution. But it would be useful to define a simple rule that people can memorize.

The rule I've been using so far is something like:

*The Dagger API is data-oriented. You query for data, and dagger returns that data. If executing a command is necessary to return that data, it will do so. When crafting a pipeline, ask yourself: what data do you need that pipeline to return? Is it an artifact? The result or standard output of a command? Most of the time, actually executing a command is just an implementation detail of what you actually want*